### ISSUE TITLE
TString copy constructor does not use the correct allocator pool (possibly other types as well)

### ISSUE THREAD

**Issue by mbechard** [KhronosGroup/glslang#2760]
If you construct a TString via
`TString s(someOtherTString)`
It incorrectly uses the allocator pool of someOtherTString, instead of the current pool for the thread. This is a problem if the pool ends up getting deleted before the string does. This causes simple workflows to be error-prone, since you need to construct your TStrings using the following code, to make it use the correct pool.
```
TString s;
s = someOtherTString;
```

Ran into this during some changes to `SetupBuiltinSymbolTable()`, where it's trying to create copies of objects allocated in one pool using a different pool.

**Comment by greg-lunarg** [comment]
TString is really just a typedef of basic_string defined in MS's xstring header file, so what we are seeing is a consequence of that, and we will not be able to fix the problem directly by changing the basic_string copy constructor since we cannot change xstring.

It is seeming to me it will take a little reworking of TString to get this right. We would probably have to make TString a class of it's own to be able to the right thing for the original code above.

**Comment by greg-lunarg** [comment]
Pretty sure I found that basic_string copy constructor does indeed use the allocator of the string to be copied :(

**Comment by greg-lunarg** [comment]
Another option is to re-evaluate the original code. Is it really good practice for one thread to copy from another's thread-local memory? Is there a better way to do this?

**Comment by greg-lunarg** [comment]
That last question was rhetorical :)

**Comment by mbechard** [comment]
Well it's not really different threads here, just different pools being used within one thread.

**Comment by jeremy-lunarg** [comment]
I believe we _could_ implement this behavior using `allocator_traits::select_on_container_copy_construction`. However, I'm not familiar enough with the allocation strategy to know whether we _should_.

Does `NewPoolTString` help? e.g. `TString s(NewPoolTString(someOtherTString.c_str()));`

**Comment by mbechard** [comment]
I do wonder if these pools are even needed. Aren't modern runtime's multi-threaded allocators fast enough for these usage cases? If certain object types are allocated/deallocated so often that they are a speed issue, they could be put onto recycle lists as an alternative.

**Comment by mbechard** [comment]
Yeah, just hitting this more today when trying to clean up resources. It's not just a thread issue, every TShader creates it's own pool. Then if you try to copy information from those shaders into other places that information becomes invalid if you delete the shader.
For example I need to implement the `TCall` copy constructor manually to avoid the TString copy constructor, and `TIntermediate::mergeTrees` needs to manually copy the TStrings between `ioAccessed` sets to ensure the newly created strings are using the correct allocator.

**Comment by jeremy-lunarg** [comment]
@mbechard 
Can you try this patch?

https://github.com/jeremy-lunarg/glslang/commit/f02685c1ce7a00c406c7b8d72a4c9aea495bdcc7

**Comment by mbechard** [comment]
Thanks Jeremy, I'll be back at my machine next week that has those test cases and I'll give this a try.
